1. Importing Libraries

In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

2. Importing Dataframes

In [4]:
# Create String of Path to Main Project Folder
path = r'/Users/seema/Documents/CareerFoundry/A4/Instacart Basket Analysis'

In [5]:
# Import Dataframes
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

3. Clean df_prods

In [6]:
# Create a Dataframe & Add Mixed Type Column
df_test = pd.DataFrame()
df_test['mix'] = ['a','b',1,True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Check for Mixed Types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
# Convert 'mix' Data Type from Numeric to String
df_test['mix'] = df_test['mix'].astype('str')

In [11]:
# Find Missing Values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# Show Null Values
df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
df_prods.shape

(49693, 5)

In [14]:
# Create Clean Dataframe without Missing Values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]
df_prods_clean.shape

(49677, 5)

In [15]:
# Find Full Duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [16]:
df_prods_clean.shape

(49677, 5)

In [17]:
# Drop Duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()
df_prods_clean_no_dups.shape

(49672, 5)

In [18]:
# Export df_prods_clean_no_dups
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

4. Clean df_ords

In [19]:
df_ords.describe()

Unnamed: 0      order_id       user_id      orders_#  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

Q2: Many of the values look reasonable, for days of the week the max is 6 and min is 0, which represents the 7 days, for hours I would expect the min to be 0 and the max to be 23, days since prior order are all positive values, which makes, sense, and the other values are only identifiers and cannot be analyzed the same way as the other values.

In [25]:
# Check for Mixed Types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

Q4: I did not find any mixed type data

In [26]:
# Find Missing Values df_ords
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
orders_#                       0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Q5: There are 206,209 missing values for days_since_prior_order. This could be because of a lack of a prior order, it's possible that these values are not applicable.
Q6: Given that values are likely missing and are correctly missing, I will not adjust the data.

In [28]:
# Identify Duplicates in df_ords
df_dups = df_ords[df_ords.duplicated()]

In [29]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, orders_#, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

Q8: There are no duplicates found in the new subset df_dups

In [30]:
# Export df_ords
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))